### ISSUE TITLE
RectMapCollider, player sometimes stuck

### ISSUE THREAD

**Issue by ccanepa** [los-cocos/cocos#248]
initial report by Netanel at
https://groups.google.com/forum/#!topic/cocos-discuss/a494vcH-u3I

The defect is that the player gets stuck at some positions.

**Comment by ccanepa** [comment]
A bugdemo checked in at https://github.com/los-cocos/etc_code/tree/master/cocos%23248--RectMapCollider%2C%20player%20sometimes%20stuck

It is a modification on Netadel bugdemo; needs blinker (available on pypi) to run.

Issue confirmed for cocos master Aug 1, 2015 (292ae676) and cocos-0.6.3-release

Further investigation shows that this happens when both of this concur
* the player actively pushes against a blocking surface
* player rect aligns with the grid tile.

**Comment by ccanepa** [comment]
Added a partial workaround with commit 3bab49139d2

Probably usable in the OP case but still seen two stuck cases
* player pushes a wall on the left at the same time it wants to go down
* player pushes a wall on the top side and wants to go left

It only stuck at grid crossings

The stuck in non definitive: releasing the pressure against the wall resumes normal
  
If not actively pushing against the wall the movement is normal

**Comment by ccanepa** [comment]
theres a third stuck case:
* player pushes a wall on the right and wants to go down

**Comment by dangillet** [comment]
I unfortunately don't have much time. Here is quickly what the problem is. If your last rect is on the ground, and its left side just aligned with the left side of the tile below. If you try to go down and left, when doing do_collision, you will go through the tile at the bottom left and it will detect that `new` has to be corrected for both its bottom and its left attribute.

As the code says in remarks:
>     # resolve them and re-collide if necessary; make sure the cells
>    # colliding the sprite midpoints are done first

But this is not done! Also, when doing it, you need to make sure that the corrected `new` rectangle is still colliding with the rest of the cells you have.

Here is a quick fix. As I said, I don't have much time. But it shows the way to go I think.

```python
class CustomRectMapCollider(RectMapCollider):
    """
    Same as RectMapCollider, except it publishes which cells will be considered
    for collision.

    Usage:
        # istantiate
        a = CustomRectMapCollider()
        # add the signal we want to emit
        a.signal = blinker.signal("collider cells")
        # use as stock RectMapCollider
        # catch the signal with something like ShowCollision
    """
    def collide_map(self, map, last, new, dy, dx):
        """same as original except the self.signal... line"""
        self.resting = False
        tested = set()
        cells = map.get_in_region(*(new.bottomleft + new.topright))
        def midpoints_first(cell):
            for midpoint in (new.midright, new.midleft, new.midtop, new.midbottom):
                if cell.contains(*midpoint):
                    return True
            return False
        cells.sort(key=midpoints_first, reverse=True)
        # next line added for debug
        self.signal.send(payload=cells)
        for cell in cells:
            if cell is None or cell.tile is None or not cell.intersects(new):
                continue
            # don't re-test
            key = (cell.i, cell.j)
            if key in tested:
                continue
            tested.add(cell)
            dx, dy = self.do_collision(cell, last, new, dy, dx)
        return dx, dy
```
And change in *rect.py*
```python
    def intersects(self, other):
        """Return boolean whether the "other" rect (an object with .x, .y,
        .width and .height attributes) overlaps this Rect in any way.
        """
        if self.x + self.width <= other.x:
            return False
        if other.x + other.width <= self.x:
            return False
        if self.y + self.height <= other.y:
            return False
        if other.y + other.height <= self.y:
            return False
        return True
```

**Comment by ccanepa** [comment]
hey, that was fast!

At a quick glance, your suggestion seems more correct that current code. 

Still, I think they are edge cases not covered, by example if player width/size are 2x the tile width, and we push against a wall with 1-tile holes so that midpoint hits a hole: the ordering should do nothing, thus decaying to previous behavior.

I will look later. Appreciate your contributions.

**Comment by dangillet** [comment]
I made a pull request for you to look at. It's a different method. I don't know if it's *the* solution.

I also noticed that `TmxObjectMapCollider` was doing almost the same as `RectMapCollider`. Actually there was even a bug in `RectMapCollider`, as the `tested` set was filled with `cell` instances, but we check if a tuple `(cell.i, cell.j)` was in the set, which would never happen.

What should be done with the  `TmxObjectMapCollider` ?

Daniel

**Comment by ccanepa** [comment]
I will look at this at local tonight, thanks for caring.

**Comment by ccanepa** [comment]
at commit 7d9728d95 to master added some utest and visualizations for this issue.

With current cocos got 30 failed, 110 passed , some failures are new [but needs review, can be a problem in the test]

The tests need  more work to enable another bunch of parametrized tests.

I not tested with the code in your PR, too late today.

Feel free to review  or update the tests; I doubt this week will have much time to put here.

Also, given that you are active, do you want commit rights ?  It simplifies a bit the branching and merging.

**Comment by dangillet** [comment]
Hi Claudio,

These tests are really terrific! :) It took me a while to go through the whole process, but I finally grasped them.

I tried running them with my code, and I got 22 fails. Looking deeper I realized I had to make a small adjustment to my code. I made a new commit on my PR https://github.com/los-cocos/cocos/pull/249 and I only now have 2 fails. I think these 2 tests are not correct. So please see my [last comment](https://github.com/los-cocos/cocos/pull/249#issuecomment-144010294) on my PR.

Regarding the commit rights, I feel honoured. :) I just want to let you know that I'm only programming as a hobby. I'm by no mean a professional and I don't have a CS degree or something. But I'm eager to learn.

**Comment by ccanepa** [comment]
(PR addressed in the PR thread)
>Regarding the commit rights, I feel honoured. :) I just want to let you know that I'm only programming as a hobby.

Fair warning. Don't worry about that.
I sent you an invite, so after accepting you should have commit rights.
For docstrings or trivial fixes like typos you can commit directly to master, for new features or changes in behavior it would be better in a branch on the los-cocos/cocos repo

Wellcome!

**Comment by dangillet** [comment]
So after quite some time trying to properly fix this issue, it is now fixed by commit fcf5fd0bf.
